src/DataLib/fendl32B_retrofit/groupr_tools.py

```diff
@@ -0,0 +1,210 @@
+    
+    # Construct the filetype and isotope specific URL
+    isotope_component = f'{element}/{element}{A}/lib/endf/n-{element}{A}.'
+    ext = file_handling[filetype.lower()]['ext']
+    download_url = tendl_gen_url + isotope_component + ext
+
+    # Define a save path for the file if there is not one already specified
+    if save_path is None:
+        save_path = f'tape{file_handling[filetype.lower()]["tape_num"]}'
+
+    # Check if the file exists
+    try:
+        urllib.request.urlopen(download_url)
+    except urllib.error.URLError as e:
+        file_not_found_code = 404
+        if str(file_not_found_code) in str(e):
+            raise FileNotFoundError()
+
+    # Download the file using urllib
+    with urllib.request.urlopen(download_url) as f:
+        temp_file = f.read().decode('utf-8')
+    
+    # Write out the file to the save_path
+    with open(save_path, 'w') as f:
+        f.write(temp_file)
+
+    return save_path
+
+# Define a function to redirect special ENDFtk output to logger
+@contextlib.contextmanager
+def redirect_ENDFtk_output():
+    # Redirect stdout and stderr to logger
+    logger_stdout = LoggerWriter(logger.info)
+    logger_stderr = LoggerWriter(logger.error)
+
+    with open(os.devnull, 'w') as fnull:
+        old_stdout = os.dup(1)
+        old_stderr = os.dup(2)
+        # Suppress terminal stdout readout
+        os.dup2(fnull.fileno(), 1)
+
+        sys.stdout = logger_stdout
+        sys.stderr = logger_stderr
+        try:
+            yield
+        finally:
+            os.dup2(old_stdout, 1)
+            os.dup2(old_stderr, 2)
+            os.close(old_stdout)
+            os.close(old_stderr)
+            # Reset stdout and stderr to default
+            sys.stdout = sys.__stdout__
+            sys.stderr = sys.__stderr__
+
+# Define a function to extract MT and MAT data from an ENDF file
+def endf_specs(endf_path):
+    with redirect_ENDFtk_output():
+        # Read in ENDF tape using ENDFtk
+        tape = ENDFtk.tree.Tape.from_file(endf_path)
+
+        # Determine the material ID
+        mat_ids = tape.material_numbers
+        matb = mat_ids[0]
+
+        # Set MF for cross sections
+        xs_MF = 3
+
+        # Extract out the file
+        file = tape.material(matb).file(xs_MF).parse()
+
+        # Extract the MT numbers that are present in the file
+        MTs = [MT.MT for MT in file.sections.to_list()]
+
+    return matb, MTs
+
+# Define a function to format GROUPR input cards
+def format_card(card_name, card_content, MTs):
+    card_str = ''
+    gen_str = ' ' + ' '.join(map(str, card_content))
+    if card_name == 9:
+        card_str = ' ' + '/\n '.join(card_content) + '/\n'
+    elif card_name == 4:
+        card_str += gen_str + '\n'
+    else:
+        card_str += gen_str + '/\n'
+    return card_str
+
+# Define a function to create the GROUPR input file
+def groupr_input_file_format(matb, MTs, element, A, mt_table):
```

Comment:
List out these constants at the top the file and then just call them in the function. All CAPS format for constants per Micah.